File: java/src/main/java/benchmark/serializers/TypeUtil.java
package benchmark.serializers;

import benchmark.model.v2.Document;
import benchmark.model.v2.Event;
import benchmark.model.v2.Message;
import benchmark.model.v2.Strings;
import benchmark.model.v2.Telemetry;
import com.fasterxml.jackson.core.type.TypeReference;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

/** Helpers for typed empty targets and list TypeReferences. */
public final class TypeUtil {
  private TypeUtil() {}

  public static Class<?> elementClass(Object value) {
    if (value instanceof List<?> list && !list.isEmpty()) {
      return list.get(0).getClass();
    }
    return value.getClass();
  }

  public static boolean isList(Object value) {
    return value instanceof List<?>;
  }

  @SuppressWarnings("unchecked")
  public static <T> T newEmpty(Object prototype) {
    if (prototype instanceof List<?> list) {
      // Return empty ArrayList; serializers that need typed list use TypeReference.
      return (T) new ArrayList<>();
    }
    try {
      return (T) prototype.getClass().getDeclaredConstructor().newInstance();
    } catch (ReflectiveOperationException e) {
      throw new IllegalStateException("cannot instantiate " + prototype.getClass(), e);
    }
  }

  public static TypeReference<?> listTypeRef(Object prototype) {
    if (!(prototype instanceof List<?> list) || list.isEmpty()) {
      throw new IllegalArgumentException("expected non-empty List prototype");
    }
    Object first = list.get(0);
    if (first instanceof Message) return new TypeReference<List<Message>>() {};
    if (first instanceof Document) return new TypeReference<List<Document>>() {};
    if (first instanceof Telemetry) return new TypeReference<List<Telemetry>>() {};
    if (first instanceof Strings) return new TypeReference<List<Strings>>() {};
    if (first instanceof Event) return new TypeReference<List<Event>>() {};
    return new TypeReference<List<Object>>() {};
  }

  public static Class<?> javaType(Object prototype) {
    if (prototype instanceof List<?>) {
      return List.class;
    }
    return prototype.getClass();
  }

  /** Create an empty array of the prototype element type (for codecs that use arrays). */
  public static Object emptyArrayLike(Object prototype) {
    Class<?> el = elementClass(prototype);
    return Array.newInstance(el, 0);
  }
}

File: kotlin/src/main/kotlin/benchmark/serializers/TypeUtil.kt
package benchmark.serializers

import benchmark.model.v2.Document
import benchmark.model.v2.Event
import benchmark.model.v2.Message
import benchmark.model.v2.Strings
import benchmark.model.v2.Telemetry
import com.fasterxml.jackson.core.type.TypeReference
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.serializer

/** Helpers for typed empty targets, list TypeReferences, and kotlinx serializers. */
object TypeUtil {
    fun isList(value: Any): Boolean = value is List<*>

    fun elementClass(value: Any): Class<*> {
        if (value is List<*> && value.isNotEmpty()) {
            return value[0]!!.javaClass
        }
        return value.javaClass
    }

    fun listTypeRef(prototype: Any): TypeReference<*> {
        val list = prototype as? List<*> ?: throw IllegalArgumentException("expected List prototype")
        if (list.isEmpty()) throw IllegalArgumentException("expected non-empty List prototype")
        return when (list[0]) {
            is Message -> object : TypeReference<List<Message>>() {}
            is Document -> object : TypeReference<List<Document>>() {}
            is Telemetry -> object : TypeReference<List<Telemetry>>() {}
            is Strings -> object : TypeReference<List<Strings>>() {}
            is Event -> object : TypeReference<List<Event>>() {}
            else -> object : TypeReference<List<Any>>() {}
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun kotlinxSerializer(value: Any): KSerializer<Any> {
        if (value is List<*>) {
            val first = value.firstOrNull() ?: throw IllegalArgumentException("empty list")
            val el = elementSerializer(first)
            return ListSerializer(el) as KSerializer<Any>
        }
        return elementSerializer(value) as KSerializer<Any>
    }

    @Suppress("UNCHECKED_CAST")
    fun elementSerializer(value: Any): KSerializer<Any> {
        return when (value) {
            is Message -> Message.serializer() as KSerializer<Any>
            is Document -> Document.serializer() as KSerializer<Any>
            is Telemetry -> Telemetry.serializer() as KSerializer<Any>
            is Strings -> Strings.serializer() as KSerializer<Any>
            is Event -> Event.serializer() as KSerializer<Any>
            else -> serializer(value::class.java) as KSerializer<Any>
        }
    }
}
